keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -132,48 +133,97 @@ Other identified non-goals are:
+`Status.ContainerStatuses[i].ResourcesAllocated` when considering available
+space on a node.
+
+#### Subresource
 
-Additionally, Kubelet is authorized to update PodSpec, and NodeRestriction
-admission plugin is extended to limit Kubelet's update access only to Pod's
-ResourcesAllocated field for CPU and memory resources.
+For alpha, resource changes will be made by updating the pod spec.  For beta
+(or maybe a followup in alpha), a new subresource, /resize, will be defined.
+This subresource could apply to other resources that carry PodTemplates, such
+as Deployments, ReplicaSets, Jobs, and StatefulSets.  Thie will allow users to
+grant RBAC access to controllers like VPA without allowing full write access to
+pod specs.
+
+The exact API here is TBD.
 
 #### Container Resize Policy
 
 To provide fine-grained user control, PodSpec.Containers is extended with
-ResizePolicy - a list of named subobjects (new object) that supports 'cpu'
-and 'memory' as names. It supports the following policy values:
-* NoRestart - the default value; resize Container without restarting it,
-* RestartContainer - restart the Container in-place to apply new resource
-  values. (e.g. Java process needs to change its Xmx flag)
-
-By using ResizePolicy, user can mark Containers as safe (or unsafe) for
-in-place resource update. Kubelet uses it to determine the required action.
+ResizePolicy - a list of named subobjects (new object) that supports 'cpu' and
+'memory' as names. It supports the following policy values:
+* RestartNotRequired - default value; try to resize the Container without
+  restarting it, if possible.
+* Restart - the container requires a restart to apply new resource values.
+  (e.g.  Java process needs to change its Xmx flag) By using ResizePolicy, user
+  can mark Containers as safe (or unsafe) for in-place resource update. Kubelet
+  uses it to determine the required action.
+
+Note: `RestartNotRequired` does not guarantee that container won't be
+restarted. The runtime may choose to stop the container if it is unable to
+apply the new resources without doing so.
 
 Setting the flag to separately control CPU & memory is due to an observation
 that usually CPU can be added/removed without much problem whereas changes to
 available memory are more probable to require restarts.
 
-If more than one resource type with different policies are updated, then
-RestartContainer policy takes precedence over NoRestart policy.
+If more than one resource type with different policies are updated at the same
+time, then any `Restart` policy takes precedence over `RestartNotRequired` policies.
+
+If a container's RestartPolicy is `Never`, the ResizePolicy must be set to
+`RestartNotRequired` to pass validation.  That said, any in-place resize may
+result in the container being stopped *and not restarted*, if the system can
+not perform the resize in place.
+
+#### Resize Status
+
+In addition to the above, a new field `Pod.Status.ContainerStatuses[i].Resize`
```

Comment:
Do we need this at per-container level? When we get a pod resize request, we consider the summation of container resize request as an atomic unit. It is all or none deal just as when scheduling a new pod.

Reply:
>That seems OK. It's not ambiguous. Now what happens if there are 2 roughly simultaneous change to the 2 containers? What should happen?
>
> What if only 1 is feasible - should kubelet actuate that? What value goes here in that case? Now it's ambiguous. This feels like a stupid corner case but I think we need to spec it.
> 
> So I still think having this (or something like it) per-container removes ambiguity. Is there a reason to hide the details?

In the API, we have status.containerStatuses[i].ResourcesAllocated that allows us communicate at per-container granularity of what requests are accepted at any given point.

Let's expand your example above, intentionally making it a bit more complex (and use ‘Proposed’ instead of ‘Unknown’):
```
T=0: New pod enters running state
    spec.containers[0].resources.requests[cpu] = 1
    status.containerStatuses[0].resourcesAllocated[cpu] = 1
    status.Resize[cpu] = ''  (Initial, no value)
    status.containerStatuses[0].resources.requests[cpu] = 1
T=1: User update #1 requesting cpu = 1.5. apiserver patches pod object, sets Resize to ‘Proposed’.
    spec.containers[0].resources.requests[cpu] = 1.5
    status.containerStatuses[0].resourcesAllocated[cpu] = 1
    status.Resize[cpu] = 'Proposed'   <-- apiserver sets this because user changed spec resources via /resize subresource.   
    status.containerStatuses[0].resources.requests[cpu] = 1
T=2: Kubelet watching the pod sees update #1, accepts it, sends patch {status.containerStatuses[0].resourcesAllocated[cpu] = 1.5}, and then invokes CRI
    spec.containers[0].resources.requests[cpu] = 1.5
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'InProgress'   <-- Note: apiserver, not kubelet, sets ‘InProgress’ when {foreach c spec.requests[cpu] == status.resourcesAllocated[cpu] != status.resources[cpu]} is true
    status.containerStatuses[0].resources.requests[cpu] = 1
T=3: User update #2 requesting cpu = 2, apiserver patches pod object, sets Resize to ‘Proposed’.
    spec.containers[0].resources.requests[cpu] = 2
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'Proposed'   <-- apiserver sets this because user changed spec resources via /resize 
    status.containerStatuses[0].resources.requests[cpu] = 1
T=4: Container runtime applied cpu=1.5, CRI returns success, kubelet patches status.resources[cpu] = 1.5
    spec.containers[0].resources.requests[cpu] = 2
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'Proposed’   <-- Stays at ‘Proposed’ as {foreach atleast one container spec.requests[cpu] != status.resourcesAllocated[cpu]}
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=5: No-op
T=6: Kubelet sees update #2, possible but not right now, maybe later, sends patch {spec.containers[0].resources.requests[cpu] = 2, status.Resize = Deferred}
    spec.containers[0].resources.requests[cpu] = 2
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'Deferred' <-- apiserver accepts transition from Proposed to Deferred (Note: If not ever possible, this would be ‘Infeasible’)
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=7: User sees 'Deferred' , and user policy allows user to wait for pod resize at a later time, so user waits.
T=8: Some other pod leaves, kubelet can now accommodate update #2, so it sends patch {status.containerStatuses[0].resourcesAllocated[cpu] = 2}, and then invokes CRI
    spec.containers[0].resources.requests[cpu] = 2
    status.containerStatuses[0].resourcesAllocated[cpu] = 2
    status.Resize[cpu] = 'InProgress'   <-- apiserver sets status.Resize=‘InProgress’ when {foreach c spec.requests[cpu] == status.resourcesAllocated[cpu] != status.resources[cpu]} is true.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=9: Container runtime applied cpu=2, CRI returns success, kubelet patches status.resources[cpu]    
    spec.containers[0].resources.requests[cpu] = 2
    status.containerStatuses[0].resourcesAllocated[cpu] = 2
    status.Resize[cpu] = ''   <-- apiserver clears this because foreach c spec.requests[cpu] == status.resourcesAllocated[cpu] == status.resources[cpu]
    status.containerStatuses[0].resources.requests[cpu] = 2
```

Above, if at T=5 a new request came in, say requesting cpu=1.6 which kubelet can accommodate, then apiserver won’t apply the Deferred status at T=6 because Deferred patch from kubelet is qualified with spec.containers[0].resources.requests[cpu] = 2. The above sequence would then proceed as below:
```
...
...
T=5: User update #3 requesting cpu = 1.6, apiserver patches pod
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'Proposed'   <-- apiserver sets this because user changed spec resources (cpu 2 -> 1.6) via /resize subresource.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=6: Kubelet sees update #2 (request.cpu = 2), possible but not right now, sends patch {spec.containers[0].resources.requests[cpu] = 2, status.Resize = Deferred}
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = ‘Proposed’   <-- apiserver drops transition from Proposed to Deferred because Deferred applies to cpu = 2, not the latest proposed 1.6
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=7: Kubelet sees update #3, accepts it, sends patch {status.containerStatuses[0].resourcesAllocated[cpu] = 1.6}, and then invokes CRI
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.6
    status.Resize[cpu] = 'InProgress'   <-- apiserver sets ‘InProgress’ when {foreach c spec.requests[cpu] == status.resourcesAllocated[cpu] != status.resources[cpu]} is true
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=8: Container runtime applied cpu=1.6, CRI returns success, kubelet patches status.resources[cpu] = 1.6
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.6
    status.Resize[cpu] = ''   <-- apiserver clears this because foreach c spec.requests[cpu] == status.resourcesAllocated[cpu] == status.resources[cpu]
    status.containerStatuses[0].resources.requests[cpu] = 1.6
```

Expanding further, if multiple containers are modified in a request, the above would still work because the patch from kubelet either atomically accepts the change to all containers, or rejects them all as a unit with Deferred/Infeasible response by specifying the entire evaluated change that was rejected.

From the user perspective, to me it seems simpler to consume summarized Proposed/InProgress/Deferred/InFeasible applicable to all Requests at any given time vs. walking an array of these Proposed/InProgress/Deferred/InFeasible statuses and trying to figure out what’s really going on. What do you think?

Please look through the above and see if I missed anything. I personally would have more confidence once I prototyped it. I hope to get to it in the next few weeks once I complete porting the earlier implementation to latest code and fixing it up for updated design (kubelet checkpointing being the main new thing).

I plan to get started on the coding in a week or so after clearing up my day-job backlog. I do want to hammer this feature out in the upcoming release (at least he core parts per updated design). This last unresolved item of signaling the user could come in beta if I’m short of time.

Reply:
I think you made a mistake.  I'm going to re-flow a bit for readability.

You said in the second code block:


```
T=5: User update #3 requesting cpu = 1.6, apiserver patches pod
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = 'Proposed'
      ^^ apiserver sets this because user changed spec resources
         (cpu 2 -> 1.6) via /resize subresource.
    status.containerStatuses[0].resources.requests[cpu] = 1.5
T=6: Kubelet sees update #2 (request.cpu = 2), possible but not right now
    kubelet sends patch {
        spec.containers[0].resources.requests[cpu] = 2,
        status.Resize = Deferred
    }
    spec.containers[0].resources.requests[cpu] = 1.6
    status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    status.Resize[cpu] = ‘Proposed’cw
      ^^ apiserver drops transition from Proposed to Deferred because
         Deferred applies to cpu = 2, not the latest proposed 1.6
    status.containerStatuses[0].resources.requests[cpu] = 1.5
```

After the patch a T=6, `spec.containers[0].resources.requests[cpu]` would be 2.
This isn't a precondition, it's a patch.  API server doesn't know that it
should not set the value to 2.0 unless we handle that specially (and let's not
open that can of worms).  I am not sure this fundamentally changes what you are
saying, though.  For the remainder of the example, a few things change but the
end result is the same.

What I don't really care for is the automatic summarization in `Resize`.  I'm
going to copy your merged code blocks and change it and call out my diffs.
the big cross-cutting diff is that Resize is not automatic.

```
T=0: New pod enters running state
    - spec.containers[0].resources.requests[cpu] = 1
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1
    - status.Resize[cpu] = ''  (Initial, no value)
    - status.containerStatuses[0].resources.requests[cpu] = 1

T=1: User update #1 requesting cpu = 1.5.
    - apiserver patches pod, sets Resize to ‘Proposed’.
    - spec.containers[0].resources.requests[cpu] = 1.5
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1
    - status.Resize[cpu] = 'Proposed'
       ^^ apiserver sets this because user changed spec resources via /resize.
    - status.containerStatuses[0].resources.requests[cpu] = 1

T=2: Kubelet watching the pod sees update #1, accepts it,
    - sends patch {
        status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
        status.Resize[cpu] = 'InProgress'
      }
    - spec.containers[0].resources.requests[cpu] = 1.5
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    - status.Resize[cpu] = 'InProgress'
    - status.containerStatuses[0].resources.requests[cpu] = 1

T=3: User update #2 requesting cpu = 2.
    - apiserver patches pod, sets Resize to ‘Proposed’.
    - spec.containers[0].resources.requests[cpu] = 2
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    - status.Resize[cpu] = 'Proposed'
       ^^ apiserver sets this because user changed spec resources via /resize.
    - status.containerStatuses[0].resources.requests[cpu] = 1

T=4: Container runtime applied cpu=1.5
    - kubelet sends patch {
        status.resources[cpu] = 1.5
        status.Resize[cpu] = ''
      }
```

Here's where we hit the first bump.  In order to not blow away `Resize`, this
patch must be sent with a resourceVersion, so the patch will fail.

Now, carrying on:

```
T=5: kubelet refreshes pod and sees update #2 (request.cpu = 2), possible but not right now
    - sends patch {
        status.resources[cpu] = 1.5
        status.Resize[cpu] = 'Deferred'
      }
    - spec.containers[0].resources.requests[cpu] = 2
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    - status.Resize[cpu] = 'Deferred'
    - status.containerStatuses[0].resources.requests[cpu] = 1.5

T=6: User update #3 requesting cpu = 1.6, apiserver patches pod
    - apiserver patches pod, sets Resize to ‘Proposed’.
    - spec.containers[0].resources.requests[cpu] = 1.6
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.5
    - status.Resize[cpu] = 'Proposed'
       ^^ apiserver sets this because user changed spec resources via /resize.
    - status.containerStatuses[0].resources.requests[cpu] = 1.5

T=7: Kubelet sees update #3, accepts it,
    - sends patch {
        status.containerStatuses[0].resourcesAllocated[cpu] = 1.6
        status.Resize[cpu] = 'InProgress'
      }
    - spec.containers[0].resources.requests[cpu] = 1.6
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.6
    - status.Resize[cpu] = 'InProgress'
    - status.containerStatuses[0].resources.requests[cpu] = 1.5

T=8: Container runtime applied cpu=1.6
    - kubelet sends patch {
        status.resources[cpu] = 1.6
        status.Resize[cpu] = ''
      }
    - spec.containers[0].resources.requests[cpu] = 1.6
    - status.containerStatuses[0].resourcesAllocated[cpu] = 1.6
    - status.Resize[cpu] = ''
    - status.containerStatuses[0].resources.requests[cpu] = 1.6
```

The only real change I made is that I used resourceVersion for what it is meant
for - conflict detection.  The thing I want to draw attention to is this:
Every single time we change `resourcesAllocated` we also change `resize`.

The only time I can't infer `resourcesAllocated` is between t=3 and T=5, during
which and observer would see:
    - spec.containers[0].resources.requests[cpu] = 2
    - status.resize[cpu] = 'Proposed'
    - status.containerStatuses[0].resources.requests[cpu] = 1

An observer knows that a resize has been proposed and not acknowledged.  The
scheduler would pessimistically assume cpu=2 until T5, at which point it would
see:
    - spec.containers[0].resources.requests[cpu] = 2
    - status.resize[cpu] = 'Deferred'
    - status.containerStatuses[0].resources.requests[cpu] = 1.5

Now it is no longer ambiguous.  The scheduler could assume 1.5 if it is being
aggressive (e.g. this is the only possible node for the to-be-scheduled pod)
or 2 if it is being cautious.

In almost all cases, `resourcesAllocated` is redundant, and in the one case
it adds value, it adds very little value (IMO).

There are still corner cases to spec.  I'll try to demonstrate those more
clearly when I get back to keyboard later.  I do think that `resize` being
`status.containerStatuses[0].resize[]` is more precise.

Reply:
>After the patch a T=6, spec.containers[0].resources.requests[cpu] would be 2.
>This isn't a precondition, it's a patch. API server doesn't know that it
>should not set the value to 2.0 unless we handle that specially (and let's not
>open that can of worms).

Yes you are correct, a standard patch would undesirably be reset back to 2.0. I overloaded the 'Patch' term when I really meant special handling perhaps a via a second subresource or mutating admission controller that recognizes modification to resize field comes only from kubelet and can correctly reconcile competing modifications. (User/VPA never modify resize field this as they go via /resize subresource.)

I agree we can do away with resourcesAllocated in the API (and if we really need it, can add back later easily). So let's continue discussing this in the below thread [without resourcesAllocated in the picture](https://github.com/kubernetes/enhancements/pull/1883#discussion_r555668219).